### DISCUSSION TITLE
question on anyon hubbard model

### DISCUSSION THREAD

**Discussion by messicr458** [QuSpin/QuSpin#530] [General]
We use fractional Jordan-winger transformation to simulate anyon Hubbard model.
The code is as following:
from quspin.operators import hamiltonian # Hamiltonians and operators
from quspin.basis import boson_basis_1d # Hilbert space boson basis
import numpy as np # generic math functions
##### define model parameters #####
L=6 # system size
J=1.0 # hopping
U=1 # interaction
mu=-1.0 # chemical potential
theta=np.pi/2
basis = boson_basis_1d(L,Nb=L) # full boson basis
#print(basis)
Jlist=[[-J*np.exp(1j*theta),i,(i+1)] for i in range(L)] #PBC
Ulist=[[0.5*U,i,i] for i in range(L)] # U/2 \sum_j n_j n_j
Mulist=[[-mu-0.5*U,i] for i in range(L)] # -(\mu + U/2) \sum_j j_n
Jlistconj=[[-J*np.conjugate(np.exp(1j*theta)),i,(i+1)] for i in range(L)] #PBC
#Jlistconj=np.conjugate(Jlist).T
#print(Jlist)
#print(Jlistconj)
#print(Ulist)
#print(Mulist)
static=[['+-',Jlist],['-+',Jlistconj],['n',Mulist],['nn',Ulist]]
dynamic=[]
#print(static)
H=hamiltonian(static,dynamic,basis=basis,dtype=np.float64,check_pcon=False, check_symm=False, check_herm=False)
print(H.todense())

However, the error message shows:

OpstrError                                Traceback (most recent call last)
~\AppData\Local\Temp\ipykernel_17168\3920563132.py in <module>
     22 dynamic=[]
     23 #print(static)
---> 24 H=hamiltonian(static,dynamic,basis=basis,dtype=np.float64,check_pcon=False, check_symm=False, check_herm=False)
     25 print(H.todense())

C:\Anaconda3\envs\quspin\lib\site-packages\quspin\operators\hamiltonian_core.py in __init__(self, static_list, dynamic_list, N, basis, shape, dtype, static_fmt, dynamic_fmt, copy, check_symm, check_herm, check_pcon, **basis_kwargs)
    320 
    321 
--> 322                         self._static=make_static(self._basis,static_opstr_list,dtype)
    323                         self._dynamic=make_dynamic(self._basis,dynamic_opstr_list,dtype)
    324                         self._shape = self._static.shape

C:\Anaconda3\envs\quspin\lib\site-packages\quspin\operators\_make_hamiltonian.py in make_static(basis, static_list, dtype)
    107 	"""
    108         static_list = _consolidate_static(static_list)
--> 109         return basis._make_matrix(static_list,dtype)
    110 
    111 

C:\Anaconda3\envs\quspin\lib\site-packages\quspin\basis\base.py in _make_matrix(self, op_list, dtype)
    224 
    225                 for opstr,indx,J in op_list:
--> 226                         ME,row,col = self.Op(opstr,indx,J,dtype)
    227                         if(len(ME)>0):
    228                                 imax = max(row.max(),col.max())

C:\Anaconda3\envs\quspin\lib\site-packages\quspin\basis\base.py in Op(self, opstr, indx, J, dtype)
    216 
    217 		"""
--> 218                 return self._Op(opstr,indx,J,dtype)
    219 
    220         def _make_matrix(self,op_list,dtype):

C:\Anaconda3\envs\quspin\lib\site-packages\quspin\basis\basis_1d\base_1d.py in _Op(self, opstr, indx, J, dtype)
    594                 error = self._op(row,col,ME,opstr,indx,J,*self._op_args,**self._blocks_1d)
    595 
--> 596                 if error != 0: raise OpstrError(_basis_op_errors[error])
    597 
    598                 mask = _np.logical_not(_np.logical_or(_np.isnan(ME),_np.abs(ME)==0.0))

OpstrError: attemping to use real hamiltonian with complex matrix elements.

**Comment by JiahaoYao** [comment]
Hi @messicr458 , i didn't read into your question or code. 

Could you first try change `dtype=np.float64` to `dtype=np.complex128`? 
Thanks!

**Reply by JiahaoYao** [reply]
@mgbukov @weinbe58  if this is question users frequently encounter, maybe it would be good to document the complex type error or use complex as the default type?

**Reply by messicr458** [reply]
thanks for your reply. Just delete `dtype=np.float64` is OK.

**Reply by JiahaoYao** [reply]
perfect, thanks for support `Quspin`!
